### PR TITLE
Barometric factor

### PR DESCRIPTION
**Pull request by PeterQFR**
Add a factor that handles barometric pressure readings as a constraint on the z height of a Pose3 value. 
The conversion from pressure to height is given by conversions from https://www.grc.nasa.gov/www/k-12/airplane/atmosmet.html

The model is 

measured height in meters = Pose.z() + bias + noise;

With Pose and Bias both being values in the graph. 

There are conversions both ways between pressure and height in meters. 

We are only handling the tropospheric conversion from 0 to 11km.

I have to remove CMakelists changes for installation in a Ament environment.

### CODE REVIEW

**Comment by dellaert** [comment]
Cool! Could you: 
- format with Google style, e.g. cppformat in vs code.
- Add a test with non-zero, non-identity poses
- un comment or remove other test?
Then please request a review by me.

**Comment by PeterQFR** [comment]
I've put in the non-zero tests, however I get an error with the numerical differentiation for the jacobian. Whenever there is any roll or pitch (Rx Ry) applied to the pose, the numerical derivative seems to use that to calculate the change in z in the local frame. 

I believe the analytic jacobians are correct in the evaluateError() but I would not be suprised if they are not. so would like some help deriving them if they are not correct. 

Ie at the moment, for the pose3 only Tz contributes anything to the error so the jacobian should only be dependent on it (and the bias).

**Comment by PeterQFR** [comment]
@dellaert I've applied the formatting and non-zero tests, but need some help with the Jacobians I am returning for the Pose3 value. 

It does not agree with the numerical derivative in the test when Rx or Ry is applied to the Pose3d it is like the numerical derivative is calculated in the child frame of the pose where as I am using it as per the parent frame (to borrow from ROS conventions). The analytical jacobian of the error I think is correct.

File: gtsam/navigation/BarometricFactor.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    return e != nullptr && Base::equals(*e, tol) &&
+           traits<double>::Equals(nT_, e->nT_, tol);
+}
+
+//***************************************************************************
+Vector BarometricFactor::evaluateError(const Pose3& p, const double& bias,
+                                       boost::optional<Matrix&> H,
+                                       boost::optional<Matrix&> H2) const {
+    if (H2) (*H2) = (Matrix(1, 1) << 1.0).finished();
+    if (H) (*H) = (Matrix(1, 6) << 0., 0., 0., 0., 0., 1.).finished();
```
Comment (dellaert) [inline gtsam/navigation/BarometricFactor.cpp:49]: Respect the order :-) also, this is likely to be incorrect. The translation method will actually calculate the correct Jacobian for you.

Comment (PeterQFR) [reply gtsam/navigation/BarometricFactor.cpp]: Thanks Frank, I've updated it to use the output of the Jacobian from the translation method. The jacobians have been something that I've found challenging for a while. I think its been mostly around the frames used as I typically have found it difficult to derive those given in all but the simplest of examples. The papers and tutorials I've read don't seem to go into depth on the subject. Are there any good tutorials on derivation of analytic jacobians for these type of systems?

It does seem that gtsam has a good system for dealing with this complexity, but I'd like to be better set up to handle these problems in future. 

@dellaert Thats all updated and for review.

Comment (dellaert) [reply gtsam/navigation/BarometricFactor.cpp]: Great! I started to workflows for CI. As far as the math, take a look in math.pdf, in the docs folder.

**Comment by ProfFan** [comment]
BTW @dellaert do we need a new subfolder for this factor?

**Comment by dellaert** [comment]
> BTW @dellaert do we need a new subfolder for this factor?

No, navigation is fine. Will merge now.

**Review by dellaert — APPROVED** [review]
Looks awesome! Many thanks for your contribution!

**Comment by mnissov** [comment]
Doesn't this also need an implementation for the bias dynamics? As far as I can see this factor has no consideration of the bias random walk, which I think is necessary for this to be used?

**Comment by PeterQFR** [comment]
Thats right, but that is common with a lot of factor implementations. Here the random walk is achieved by adding additional between factor between bias values at each measurement step. 
That is, you need to add an additional Between factor, linking each of your Bias Values in the graph. Its not included in this factor because that would mean this factor would then need to reference the bias of the previous timestep. 
This factor doesn't assume anything about the bias which could be constant among other things.
